### ISSUE TITLE
MenuBar

### ISSUE THREAD

**Issue by PureWeen** [dotnet/maui#4640]
### Description

Provide a cross platform API to allow the specification of a MenuBar for desktop applications.

#### IMenuBarElement
This is used to attribute a view indicating that it has a MenuBar element that needs to be taken into consideration

```C#
public interface IMenuBarElement
{
	IMenuBar? MenuBar { get; }
}
```

#### IMenuBar
Top Level container.

Alternative option here would be to use the existing [Menu](https://github.com/dotnet/maui/blob/cdf43dafaf2dd30ec8732adcb4df855465e9623f/src/Controls/src/Core/Menu.cs) type.

```C#
public interface IMenuBar : IList<IMenuBarItem>, IElement
{
}
```

#### IMenuBarItem
Used to describe each primary menu inside the `MenuBar` (File, edit, etc..)

```C#
public interface IMenuBarItem : IList<IMenuFlyoutItemBase>, IElement
{
	string Text { get; }
        MenuBarItemKey Key { get: }
}
```

### MenuBarItemKey
Used to map a MenuBarItem to a system menu

```C#
public enum MenuBarItemKey
{
    File, Edit, View, //The rest of the default menus in catalyst
}
```

#### IMenuFlyoutItem
Describes an item that the user will click on to trigger behavior

```C#
public interface IMenuFlyoutItem : IMenuFlyoutItemBase
{
	string Text { get; }
	Action Clicked { get; }
}
```

#### IMenuFlyoutSubItem
Used to create submenus

```C#
public interface IMenuFlyoutSubItem : IMenuFlyoutItemBase, IList<IMenuFlyoutItemBase>
{
	string Text { get; }
        short Position { get; } // ?? This can be used to insert sub menus at different points on catalyst
}
```

#### Platform Mappings

- WinUI : 1:1
- Catalyst
  - MenuBar => UIMenuBuilder
  - MenuBarItem => UIMenu
  - MenuFlyoutItem => UIAction/UICommand
  - MenuFlyoutSubItem => UIMenu 

#### Considerations
We currently have a [Menu](https://github.com/dotnet/maui/blob/cdf43dafaf2dd30ec8732adcb4df855465e9623f/src/Controls/src/Core/Menu.cs) API that is only used on MacOS.  This API can be used as an attached property on every single Element.  On MacOS this would become a context menu for Elements and the main menu when it was set on the Application.  Do we want to keep this one type for both scenarios? On WinUI the `FlyoutMenu's` all reuse the `MenuFlyoutItem` types but then use `MenuFlyout` as the parent element vs `MenuBarItem`/`MenuBar`

Catalyst has a default menu that it displays, so the process of interacting with the menu is more modification based instead of just creating the menu like on WinUI.  I've added a few APIs to help with this but we could do something more in v2. Just having the handlers will give users a valuable extension point to make modifications. They can just tap into the `MenuBarHandler` and from there issue commands to the builder.

### Intended Use-Case

https://developer.apple.com/documentation/uikit/uicommand/adding_menus_and_shortcuts_to_the_menu_bar_and_user_interface
https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.menubar?view=winui-3.0

**Comment by profix898** [comment]
I checked your implementation and it mostly fits my intuition (what is to say that I was able to grasp the structure right away), which I think is usually a good sign. That said, here are a few comments on the current design and naming.

1. General API design

> We currently have a Menu API that is only used on MacOS. This API can be used as an attached property on every single Element. On MacOS this would become a context menu for Elements and the main menu when it was set on the Application."

My vote is for a unification of Menu/MenuBar and MenuItem/MenuFlyoutItem (and related items).

There are two aspect to this IMO.

a) You would probably want a context menu that can be added to other elements on WinUI as well. Right now (correct me if I'm wrong), MAUI only has context menu for ListView items. Not a general concept of context menu (for other views). In WinUI there is the 'ContextFlyout' property on 'UIElement' (https://docs.microsoft.com/de-de/uwp/api/windows.ui.xaml.uielement.contextflyout), which can be used to attach a context menu to any UIElement. So, I guess this is conceptionally similar to macOS.

b) We should have one unified way of building a window-/application-level menu in MAUI. Not a type that is specific to macOS.

I understand that with Catalyst you have a default menu. But I guess for a first iteration, you could "simply" merge the MAUI menu with the pre-existing default menu (by means of using the MenuBarItemKey flag and Position property, as you suggested in your API design, although the latter might need more thinking). At this point, it would probably require some custom logic to "apply" properties, commands, etc. to the existing menu items (e.g. in File menu).

I also think that the unified design should be based on your new proposal. The current/old Menu in MAUI has this (weird) structure with _menu and _items as two separate collections. So it's used as a container (= MenuBar) and/or as a top-level menu (= MenuBarItem, e.g. File). For me this contradicts a self-documenting/explaining API design (and probably single-purpose rule). One potential reason is that its top-level menu was used (as item list) to populate the navigation menu for flyout pages (Hamburger menu).

2. Considering that menus are found in different areas of an app, I'm not sure about the current naming schema. I would suggest to rename a few things, e.g.:

MenuBar -> MenuBar is Ok

MenuBarItem -> Menu
unless you build a window-/app-level menu(bar), this will be the root item/container for context menus or simple single-level menus (e.g. Hamburger menu). For that reason, MenuBar in its name is misleading as it can be used separate of MenuBar as well.

MenuFlyoutItem -> MenuItem
MenuFlyoutSubItem -> MenuSubMenu(Item)
Why do you have both? At least on Windows (WinForm, WPF, etc.) you can have a Click event for a submenu as well. So the only difference is that a submenu contains other menu items. Other than that it has all the same properties: text, icon, accelerator, (click) events. My guess is that having two separate types makes the handler mapping unique in MAUI, right? In that case, you could still make MenuFlyoutSubItem derive from MenuFlyoutItem, not?

3. There are some features missing in your prototype: Icons, Accelerator, IsEnabled, ... as well as additional menu item types (toggle/checked, radio select, ...) Just so we keep track ... ;)

**Comment by PureWeen** [comment]
> Why do you have both? At least on Windows (WinForm, WPF, etc.) you can have a Click event for a submenu as well. So the only difference is that a submenu contains other menu items. Other than that it has all the same properties: text, icon, accelerator, (click) events. My guess is that having two separate types makes the handler mapping unique in MAUI, right? In that case, you could still make MenuFlyoutSubItem derive from MenuFlyoutItem, not?

How do you wire into a click event for `MenuFlyoutSubItem` ?  What am I missing?

![image](https://user-images.githubusercontent.com/5375137/153964133-0c97b066-9be0-4768-bb22-3f95332d5f35.png)

**Comment by PureWeen** [comment]
> I also think that the unified design should be based on your new proposal. The current/old Menu in MAUI has this (weird) structure with _menu and _items as two separate collections.

Yea, I need to get some clarity on the choices here. 

> MenuFlyoutItem -> MenuItem

Agreed

> You would probably want a context menu that can be added to other elements on WinUI as well. Right now (correct me if I'm wrong), MAUI only has context menu for ListView items. Not a general concept of context menu (for other views). 

Interestingly `Menu` is an Attached Property with `Element` so you can attach `Menu` to literally anything.  It won't do anything of value :-) but you can do it.  I do think the `Menu` structure should be reworked a little bit though.

**Comment by profix898** [comment]
> AFAICT you can't wire into a click event for MenuFlyoutSubItem ?

MenuFlyoutItem (+ RadioMenuFlyoutItem, ToggleMenuFlyoutItem, MenuFlyoutSeparator) and MenuFlyoutSubItem all derive from 
MenuFlyoutItemBase, which in turn is ultimately a UIElement. So, if you want, each one of them has all the basic events (such as Tapped, Pointer*, etc.) or even the ability to attach a RoutedEvent handler (https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.menuflyoutsubitem?view=winui-3.0). But only MenuFlyoutItem (+ RadioMenuFlyoutItem, ToggleMenuFlyoutItem) has a dedicated Click event. MenuFlyoutSubItem (+ MenuFlyoutSeparator) does not.

Looking at it again, you ARE actually mirroring the WinUI inheritance structure in your implementation. You are right. I was probably biased from WinForms or WPF!

The point where I'm currently stuck (and what makes it difficult to review implications) is this:
Your proposed implementation is pretty straight forward and contributes a set of new controls (and corresponding handlers) to MAUI. This way we can build/populate a menu bar (in code or XAML).
However, if you have `Menu` as attached property (for context menu), who "handles" the mapping to native platform? For each attached (context) menu, we would need to create the matching platform menu (Win/macOS), but obviously that's not a classical MAUI handler. And the "scope" is not limited to a small subset of elements in the visual tree (like Grid.Row/Column with GridLayout). Isn't that similar to how Behaviours (previously Effects) are designed (that you can attach additional/platform functionality to individual elements)?

```
<Image>
  <Image.Menu>
    <MenuItem Text="Item A" Clicked="ClickAHandler" ... />
    <MenuSeparatorItem />
    <MenuItem Text="Item B" ... />
  </Image.Menu>
</Image>
```

Sorry, but this is where my current MAUI knowledge ends ... You are the MAUI pro ...

**Comment by PureWeen** [comment]
> MenuFlyoutItemBase, which in turn is ultimately a UIElement. So, if you want, each one of them has all the basic events (such as Tapped, Pointer*, etc.)

Yea...... that's probably fine and does clean up a few things. The main motivation was to not opt into too much API with the initial implementation. 

Seeing as how the WinUI implementation specifically splits those two things out, they don't inherit from each other, and they only placed a Click event on the `MenuFlyoutItem` but we can do our own thing over here :-) And then I think that lets me fully remove `IMenutFlyoutBase` which I'm not really a fan of. 

> My guess is that having two separate types makes the handler mapping unique in MAUI, right? In that case, you could still make MenuFlyoutSubItem derive from MenuFlyoutItem, not?

Yea :-) I'll play with this a bit more tomorrow.  Mapping everything to handlers just makes it construct nicely and users can easily add additional handlers for parts. The other option here is a single `MenuBar` handler and then inside that handler we have hooks. 

> However, if you have Menu as attached property (for context menu), who "handles" the mapping to native platform?

I just removed `Menu` and the `Menu` attached property for now. At some later date we can probably create a type `ContextMenu` (this will map to `MenuFlyout` on WinUI) and it'll most likely be an attached property on `View` but that will need more investigation.

I've pushed up the latest set of changes that wire up most of WinUI and place a functional menu into the header.  Tomorrow we'll also see how `Catalyst` goes.

I'm also still unsure of the names 'MenuFlyoutSubItem' and `MenuFlyoutItem` but I haven't thought of any ideal substitues yet. 

`MenuSubMenuItem' and `MenuBarMenuItem` :-/
